# Stage B: Tag Along Project
### Appliances Energy Prediction

In [2]:
# import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [3]:
# load data

df = pd.read_csv('energydata_complete.csv')
pd.set_option("display.max_columns", None)
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4       RH_4         T5   RH_5        T6  \
0  19.79  44.730000  19.000000  45.566667  17.166667  55.20  7.026667   
1  19.79  44.790000  19.000000  45.992500  17.166667  55.20  6.833333   
2  19.79  44.933333  18.926667  45.890000  17.166667  55.09  6.560000   
3  19.79  45.000000  18.890000  45.723333  17.166667  55.09  6.433333   
4  19.79  45.000000  18.890000  45.530000  17.200000  55.09  6.366667   

        RH_6         T7       RH_7    T8       RH_8         T9   RH_9  \
0  84.256667  17.200000  41.626667  18.2  48.900000  17.033333  45.53   
1  84.063333  17.200000  41.560000  18.2  48.863333  17.066667  45.56   
2  83.156667  17.200000  41.433333  18.2  48.730000  17.000000  45.50   
3  83.423333  17.133333  41.290000  18.1  48.590000  17.000000  45.40   
4  84.893333  17.200000  41.230000  18.1  48.590000  17.000000  45.40   

      T_out  Press_mm_hg  RH_out  Windspeed  Visibility  Tdewpoint        rv1  \
0  6.600000        733.5    92.0   7.000000   63.000000        5.3  13.275433   
1  6.483333        733.6    92.0   6.666667   59.166667        5.2  18.606195   
2  6.366667        733.7    92.0   6.333333   55.333333        5.1  28.642668   
3  6.250000        733.8    92.0   6.000000   51.500000        5.0  45.410389   
4  6.133333        733.9    92.0   5.666667   47.666667        4.9  10.084097   

         rv2  
0  13.275433  
1  18.606195  
2  28.642668  
3  45.410389  
4  10.084097

In [4]:
# shape of data
df.shape

(19735, 29)

In [5]:
# check for missing value
df.isnull().sum()

date           0
Appliances     0
lights         0
T1             0
RH_1           0
T2             0
RH_2           0
T3             0
RH_3           0
T4             0
RH_4           0
T5             0
RH_5           0
T6             0
RH_6           0
T7             0
RH_7           0
T8             0
RH_8           0
T9             0
RH_9           0
T_out          0
Press_mm_hg    0
RH_out         0
Windspeed      0
Visibility     0
Tdewpoint      0
rv1            0
rv2            0
dtype: int64

In [6]:
# check for duplicate row
df.duplicated().any()

False

In [7]:
# summary statistics of data
df.describe()

Appliances        lights            T1          RH_1            T2  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      97.694958      3.801875     21.686571     40.259739     20.341219   
std      102.524891      7.935988      1.606066      3.979299      2.192974   
min       10.000000      0.000000     16.790000     27.023333     16.100000   
25%       50.000000      0.000000     20.760000     37.333333     18.790000   
50%       60.000000      0.000000     21.600000     39.656667     20.000000   
75%      100.000000      0.000000     22.600000     43.066667     21.500000   
max     1080.000000     70.000000     26.260000     63.360000     29.856667   

               RH_2            T3          RH_3            T4          RH_4  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      40.420420     22.267611     39.242500     20.855335     39.026904   
std        4.069813      2.006111      3.254576      2.042884      4.341321   
min       20.463333     17.200000     28.766667     15.100000     27.660000   
25%       37.900000     20.790000     36.900000     19.530000     35.530000   
50%       40.500000     22.100000     38.530000     20.666667     38.400000   
75%       43.260000     23.290000     41.760000     22.100000     42.156667   
max       56.026667     29.236000     50.163333     26.200000     51.090000   

                 T5          RH_5            T6          RH_6            T7  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      19.592106     50.949283      7.910939     54.609083     20.267106   
std        1.844623      9.022034      6.090347     31.149806      2.109993   
min       15.330000     29.815000     -6.065000      1.000000     15.390000   
25%       18.277500     45.400000      3.626667     30.025000     18.700000   
50%       19.390000     49.090000      7.300000     55.290000     20.033333   
75%       20.619643     53.663333     11.256000     83.226667     21.600000   
max       25.795000     96.321667     28.290000     99.900000     26.000000   

               RH_7            T8          RH_8            T9          RH_9  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      35.388200     22.029107     42.936165     19.485828     41.552401   
std        5.114208      1.956162      5.224361      2.014712      4.151497   
min       23.200000     16.306667     29.600000     14.890000     29.166667   
25%       31.500000     20.790000     39.066667     18.000000     38.500000   
50%       34.863333     22.100000     42.375000     19.390000     40.900000   
75%       39.000000     23.390000     46.536000     20.600000     44.338095   
max       51.400000     27.230000     58.780000     24.500000     53.326667   

              T_out   Press_mm_hg        RH_out     Windspeed    Visibility  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean       7.411665    755.522602     79.750418      4.039752     38.330834   
std        5.317409      7.399441     14.901088      2.451221     11.794719   
min       -5.000000    729.300000     24.000000      0.000000      1.000000   
25%        3.666667    750.933333     70.333333      2.000000     29.000000   
50%        6.916667    756.100000     83.666667      3.666667     40.000000   
75%       10.408333    760.933333     91.666667      5.500000     40.000000   
max       26.100000    772.300000    100.000000     14.000000     66.000000   

          Tdewpoint           rv1           rv2  
count  19735.000000  19735.000000  19735.000000  
mean       3.760707     24.988033     24.988033  
std        4.194648     14.496634     14.496634  
min       -6.600000      0.005322      0.005322  
25%        0.900000     12.497889     12.497889  
50%        3.433333     24.897653     24.897653  
75%        6.566667     37.583769     37.583769  
max       15.500000     49.996530     49.996530

### Question 12
From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two d.p.?

In [8]:
# temperature in the living room
x = df['T2']

# temperature outside the building
y = df['T6']

In [9]:
# reshape data to 2D array
data = np.array(x)
x = data.reshape(-1, 1)

In [10]:
#splitting the data into train and test set
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size=0.7, test_size = 0.3, random_state=42)

#Define the model
reg_model = LinearRegression()

# fit the model
reg_model.fit(x_train, y_train)
pred = reg_model.predict(x_test)

In [11]:
# find r2 score

from sklearn.metrics import r2_score
r2_score = r2_score(y_test, pred)
round(r2_score, 2)

0.64

#### Answer
The r squared score is 0.64.

### Question 13
Normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set. Answer the following questions:

In [12]:
# drop 'light' and 'data' variable
df = df.drop(['lights', 'date'], axis = 1)

In [13]:
# normalize the dataset using min max scaler

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
normalized_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)

features = normalized_df.drop(columns=['Appliances'])
target = normalized_df['Appliances']

In [14]:
# split the normalized data into test and train set
xn_train, xn_test, yn_train, yn_test = train_test_split(features, target, train_size=0.7, test_size = 0.3, random_state=42)

In [15]:
#Define our model
reg_model2 = LinearRegression()

# fit the model
reg_model2.fit(xn_train, yn_train)

# predict on test set
predicted_value = reg_model2.predict(xn_test)

In [16]:
# evaluating the model based on mean absolute error
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(yn_test, predicted_value)
round(mae, 2)

0.05

#### Answer
The mean absolute error is 0.05.

### Question 14
What is the Residual Sum of Squares (in two decimal places)?

In [18]:
# calculate RSS value
residual = np.sum((yn_test - predicted_value)**2)
round(residual, 2)

45.35

#### Answer
The residual sum of squares (RSS) is 45.35

### Question 15
What is the Root Mean Squared Error (in three decimal places)?

In [19]:
# find mean squared error

from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(yn_test, predicted_value))
round(rmse, 3)

0.088

#### Answer
The root mean squared error is 0.088.

### Question 16
What is the Coefficient of Determination (in two decimal places)?

In [20]:
# find r squared score

from sklearn.metrics import r2_score
r2_score = r2_score(yn_test, predicted_value)
round(r2_score, 2)

0.15

#### Answer
The coefficient of determination is 0.15

### Question 17
Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

In [21]:
# obtaining feature weights from linear model

def get_weights_df(model, feat, col_name):
    weights = pd.Series(model.coef_, feat.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', col_name]
    weights_df[col_name].round(3)
    return weights_df

In [22]:
# get weights
linear_model_weights = get_weights_df(reg_model2, xn_train, 'Linear_Model_Weight')
linear_model_weights

Features  Linear_Model_Weight
0          RH_2            -0.456698
1         T_out            -0.321860
2            T2            -0.236178
3            T9            -0.189941
4          RH_8            -0.157595
5        RH_out            -0.077671
6          RH_7            -0.044614
7          RH_9            -0.039800
8            T5            -0.015657
9            T1            -0.003281
10          rv1             0.000770
11          rv2             0.000770
12  Press_mm_hg             0.006839
13           T7             0.010319
14   Visibility             0.012307
15         RH_5             0.016006
16         RH_4             0.026386
17           T4             0.028981
18    Windspeed             0.029183
19         RH_6             0.038049
20         RH_3             0.096048
21           T8             0.101995
22    Tdewpoint             0.117758
23           T6             0.236425
24           T3             0.290627
25         RH_1             0.553547

#### Answer
From the dataframe above, RH_2 has the lowest weight while RH_1 has the highest weight.

### Question 18
Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [23]:
# training the model with ridge regularization of 0.4

from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.4)

# fit the model
ridge_reg.fit(xn_train, yn_train)

# predict on test set
ridge_pred = ridge_reg.predict(xn_test)

In [24]:
# evalaute the model using RMSE

rmse = np.sqrt(mean_squared_error(yn_test, ridge_pred))
round(rmse, 3)

0.088

#### Answer
No change is detected in the RMSE value, it still remains 0.088.

### Question 19
Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [25]:
# training the model with lasso regularisation of 0.001

from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)

# fit the model
lasso_reg.fit(xn_train, yn_train)

# predict on test set
lasso_pred = lasso_reg.predict(xn_test)

In [26]:
lasso_weights_df = get_weights_df(lasso_reg, xn_train, 'Lasso_weight')
lasso_weights_df

Features  Lasso_weight
0        RH_out     -0.049557
1          RH_8     -0.000110
2            T1      0.000000
3     Tdewpoint      0.000000
4    Visibility      0.000000
5   Press_mm_hg     -0.000000
6         T_out      0.000000
7          RH_9     -0.000000
8            T9     -0.000000
9            T8      0.000000
10         RH_7     -0.000000
11          rv1     -0.000000
12           T7     -0.000000
13           T6      0.000000
14         RH_5      0.000000
15           T5     -0.000000
16         RH_4      0.000000
17           T4     -0.000000
18         RH_3      0.000000
19           T3      0.000000
20         RH_2     -0.000000
21           T2      0.000000
22         RH_6     -0.000000
23          rv2     -0.000000
24    Windspeed      0.002912
25         RH_1      0.017880

In [27]:
# number of features with non-zero weights

non_zero_features = lasso_weights_df.query('Lasso_weight != 0')
len(non_zero_features)

4

#### Answer
Four (4) features have non-zero lasso weights.

### Question 20
What is the new RMSE with the lasso regression? (Answer should be in three (3) decimal places)

In [28]:
# root mean squared error of lasso model
rmse = np.sqrt(mean_squared_error(yn_test, lasso_pred))
round(rmse, 3)

0.094

#### Answer
The RMSE of the lasso model is 0.094.